src/OpenTelemetry.Api/Internal/ActivityHelperExtensions.cs

```diff
@@ -306,12 +323,31 @@ public static void Enumerate(ActivityEvent activityEvent, ref TState state)
                     return;
                 }
 
+                if (maxTags.HasValue)
+                {
+                    SkipAllocationFreeEnumeration(tags, ref state, maxTags.Value);
+                    return;
+                }
+
                 ActivityTagsCollectionEnumerator(
                     tags,
                     ref state,
                     ForEachTagValueCallbackRef);
             }
 
+            // TODO: When a limit has been configured an allocation-free enumerator is not used.
+            // Need to either:
+            //     1) modify the dynamically generated code to only enumerate up to the max number of items, or
+            //     2) wait until .NET 7 is released and do this more easily with the new enumerator functions
```

Comment:
If we think we'll do another stable release prior to .NET 7 then need to decide if it's worth modifying the dynamically generated allocation-free enumerator to support configured limits.

Reply:
Since this performance hit only affects users who set a limit, my proposal is to address this TODO in a follow up PR since we may be dropping support for DiagnosticSource v6.